docs: use GitHub example in readme (#479)
joerick/cibuildwheel: README.md

Example setup
-------------

To build manylinux, macOS, and Windows wheels on Github Actions, you could use this `.github/workflows/wheels.yml`:

```yaml
name: Build

on: [push, pull_request]

jobs:
  build_wheels:
    name: Build wheels on ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-20.04, windows-2019, macOS-10.15]

    steps:
      - uses: actions/checkout@v2

      - uses: actions/setup-python@v2
        name: Install Python

      - name: Install cibuildwheel
        run: python -m pip install cibuildwheel==1.7.1

      - name: Build wheels
        env:

>           CIBW_SKIP: "?p25-*" # Skip building Python 2.7 wheels on all platforms

Reviewer: There's a typo here, it should be 27, not 25. But also I don't think the setting should be written in the config like this, as users might assume that it's required for this setup to work. Also skipping 2.7 would also mandate the removal of the choco install in windows. If the line is there but commented, users can assume it's educational, not required.

Author: Oops. How about:

```suggestion
          CIBW_SKIP: "?p27-*" # Skipping Python 2.7 for this example. See docs and examples for details on Python 2.7.
```

This has several benefits:

1. It avoids having to show choco install, which is very verbose for a simple readme example and is wrong for a _large_ portion of Python 2.7 users, since it doesn't support anything above C++03 (so no pybind11, etc).
2. cibuildhwheel will need to support Python 2 as long as the  manylinux wheels have it, but we don't have to include it in the "simplest possible" example in the readme. It's already been tripping up users like in #468 
3. It can remain in the bigger example in the examples folder.

I'd hope most users are past 2.7 support, so this is the new "default", with <50% now needing the Python 2 part, making an example that skips it more useful for a minimal first example. I'll restore it if you want it back, though.

Reviewer 2: Do we really want to do this? I thought the idea was that for this example to be as minimal/general as possible?

Reviewer 3: As I see our general example does not show how to use config variables. Also, we got bug reports like #468. I`m ok with disabling python 2.7 in the initial configuration from readme and does not do this in the example yaml file.

Author: Dropping 2.7 via a variable is more minimal than adding a whole step that adds Python 2.7 but only works for C/C++<11. And does removing 2.7 from the readme example really make it less general at the end of 2020? If anything it still shows that we support it (since we remove it via a variable), and highlights that there are caveats to using it (the extra step or steps for C++11+ support).

Reviewer 2: > As I see our general example does not show how to use config variables.

There's a whole section "Options", right underneath the minimal example, so it should be reasonably clear if users read the docs?

> adding a whole step that adds Python 2.7 but only works for C/C++<11

Is that still an issue on GitHub Actions (compared to Travis)? I thought everything was much more streamlined on GHA?
But yes, if that's the case, I can probably agree.

Reviewer 2: > Is that still an issue on GitHub Actions (compared to Travis)?

Oh, I see, the installation step of the ancient Python 2.7 compiler is still there.

Reviewer 3: > There's a whole section "Options", right underneath the minimal example, so it should be reasonably clear if users read the docs?

There is rather a problem that you need to found how to set an environment variable in yaml file. 
I think that it may be unclear that cibuildwheel build wheel on all python version if the environment variable is not set. I see examples with hole os times python version matrix.

Example setup
-------------

> To build manylinux, macOS, and Windows wheels on Github and make them availble for download after the job runs, you could use this `.github/workflows/wheels.yml`:

Reviewer: minor copy tweak, more succinct-
```suggestion
To build manylinux, macOS, and Windows wheels on Github Actions, you could use this `.github/workflows/wheels.yml`:
```

```yaml
name: Build

on: [push, pull_request]

jobs:
  build_wheels:
    name: Build wheels on ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:

>         os: [ubuntu-18.04, windows-latest, macos-latest]

Reviewer 3: why `ubuntu-18.04` instead of `ubuntu-latest`?

Author: I have no idea, that's what's in the example. Mayne because `ubuntu-latest` generates a warning that it will change to `ubuntu-20.04` soon, but that example predated that warning.

I think for a 2020 example, skipping Python 2 might be useful in the example? `CIBW_SKIP: "cp27-*" # Skip building Python 2.7 on all platforms` for example?

Reviewer 3: > I think for a 2020 example, skipping Python 2 might be useful in the example? `CIBW_SKIP: "cp27-*" # Skip building Python 2.7 on all platforms` for example?

This does not skip pypy27. But I agree that short example may skip python 2.7 Also this allow to remove "Install Visual C++ for Python 2.7" step.

Reviewer: In general, I prefer to lock to specific versions when writing CI configs, so in that case, it should be`os: [ubuntu-20.04, windows-2019, macos-10.15]`. I think previously it was only possible to lock the ubuntu image, but now you can lock windows and macos too.

I also like the idea of adding a single CIBW setting.

Author: ```suggestion
        os: [ubuntu-20.04, windows-2019, macOS-10.15]
```

Assuming we want 10.15 instead of 11.0, as it's still listed as preview?

Reviewer: Yeah that's what I had assumed, because `latest` pointed to 10.15

    steps:
      - uses: actions/checkout@v2

      - uses: actions/setup-python@v2
        name: Install Python

      - name: Install cibuildwheel
        run: python -m pip install cibuildwheel==1.7.1

      - name: Install Visual C++ for Python 2.7
        if: runner.os == 'Windows'
        run: choco install vcpython27 -f -y

      - name: Build wheels

>         run: python -m cibuildwheel --output-dir wheelhouse

Reviewer: for the option, how about:
```yaml
        env:
          # add options here, for example,
          # CIBW_SKIP: "cp27-*" # Skip building CPython 2.7 on all platforms
```

>       - name: Install Visual C++ for Python 2.7
>         if: runner.os == 'Windows'
>         run: choco install vcpython27 -f -y
> 
>       - name: Build wheels
>         run: python -m cibuildwheel --output-dir wheelhouse

Author: ```suggestion
      - name: Build wheels
        env:
          CIBW_SKIP: "?p25-*" # Skip building Python 2.7 wheels on all platforms
        run: python -m cibuildwheel --output-dir wheelhouse
```
